### PR TITLE
ci: update npm before trusted publishing

### PR DESCRIPTION
**Pull request by antongulin**
## Summary
- install the latest npm CLI before the release publish step
- add an explicit npm version guard for trusted publishing support
- extend workflow regression coverage for the npm CLI prerequisite

## Root Cause
Run 26383466272 showed npm could sign provenance but still failed final publish with `E404`. npm trusted publishing docs require npm CLI >= 11.5.1 and Node >= 22.14.0; setup-node's bundled npm can lag. This makes the workflow install and verify a supported npm before publishing.

## Test Plan
- node --test test/workflow.test.mjs (red before workflow change, green after)
- npm run prepack
- npm publish --dry-run --access public --provenance

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -23,6 +23,16 @@ jobs:
           node-version: "22"
           registry-url: "https://registry.npmjs.org"
 
+      - name: Install npm with trusted publishing support
+        run: npm install -g npm@latest
+
+      - name: Verify npm trusted publishing support
+        run: |
+          set -euo pipefail
+          NPM_VERSION=$(npm --version)
```
[the remark this answers is not in the document]
Comment (antongulin) [reply .github/workflows/publish.yml:32]: Fixed in c854633: added a regression that executes the workflow version-check script against 11, 11.5, 11.5.0, 11.5.1, and 12.0.0, then hardened the guard to reject incomplete or non-finite version parts before comparing against npm >= 11.5.1.

**Comment by antongulin** [comment]
/review

Fixed review item in c854633:
- Hardened npm version parsing so incomplete versions like `11` and `11.5` fail instead of passing.
- Added focused regression coverage that executes the workflow guard directly.

Verification:
- node --test test/workflow.test.mjs
- npm run prepack
- npm publish --dry-run --access public --provenance

File: plugin/test/workflow.test.mjs [hunk 2]
```diff
@@ -28,3 +47,14 @@ test("publish workflow uses npm trusted publishing provenance", () => {
   assert.match(workflow, /run:\s*npm publish --access public --provenance\s+working-directory:\s*plugin/s)
```
[the remark this answers is not in the document]
Comment (antongulin) [reply plugin/test/workflow.test.mjs:53]: Fixed in 5ac2ddc: the guard now rejects any version part that is not all digits before parsing, so `11.5.1a` no longer passes through parseInt truncation.

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -23,6 +23,16 @@ jobs:
           node-version: "22"
           registry-url: "https://registry.npmjs.org"
 
+      - name: Install npm with trusted publishing support
+        run: npm install -g npm@latest
+
+      - name: Verify npm trusted publishing support
+        run: |
```
[the remark this answers is not in the document]
Comment (antongulin) [reply .github/workflows/publish.yml:30]: Not accepting extra parts: npm CLI versions are emitted as semver major.minor.patch in this context. The stricter guard intentionally rejects ambiguous strings like `11.5.1.0` and prerelease/suffixed values rather than attempting to normalize them for release publishing.

File: plugin/test/workflow.test.mjs [hunk 4]
```diff
@@ -28,3 +47,14 @@ test("publish workflow uses npm trusted publishing provenance", () => {
...
   assert.doesNotMatch(workflow, /NODE_AUTH_TOKEN/)
```
[the remark this answers is not in the document]
Comment (antongulin) [reply plugin/test/workflow.test.mjs:54]: Fixed in 5ac2ddc: added regression cases for `11.5.1a`, `11.5.a`, `11.5.1.0`, and empty string, alongside existing boundary cases.

**Comment by antongulin** [comment]
/review

Fixed review items in 5ac2ddc:
- Added strict digit-only validation before parsing npm version parts.
- Added invalid-input coverage for suffixes, non-numeric patch, extra part, and empty version.
- Kept extra-part/prerelease rejection intentional for this release guard.

Verification:
- node --test test/workflow.test.mjs
- npm run prepack
- npm publish --dry-run --access public --provenance

File: plugin/test/workflow.test.mjs [hunk 5]
```diff
@@ -19,6 +35,9 @@ test("publish workflow uses npm trusted publishing provenance", () => {
   const workflow = readFileSync(publishWorkflowPath, "utf-8")
 
   assert.match(workflow, /permissions:\s+contents:\s*write\s+id-token:\s*write/s)
+  assert.match(workflow, /run:\s*npm install -g npm@latest/)
+  assert.match(workflow, /NPM_VERSION=\$\(npm --version\)/)
+  assert.match(workflow, /too old for trusted publishing/)
```
[the remark this answers is not in the document]
Comment (antongulin) [reply plugin/test/workflow.test.mjs:46]: Fixed in 8914ab2: added a 1s timeout to the child process used by the workflow guard test helper.

File: plugin/test/workflow.test.mjs [hunk 6]
```diff
@@ -19,6 +35,9 @@ test("publish workflow uses npm trusted publishing provenance", () => {
   const workflow = readFileSync(publishWorkflowPath, "utf-8")
```
[the remark this answers is not in the document]
Comment (antongulin) [reply plugin/test/workflow.test.mjs:41]: Fixed in 8914ab2: the helper now finds the workflow line containing both `node -e` and `$NPM_VERSION` before extracting the quoted script, which keeps the regex focused on the selected line.

**Comment by antongulin** [comment]
/review

Fixed review items in 8914ab2:
- Added a timeout to the workflow guard child-process test helper.
- Narrowed script extraction to the workflow line containing `node -e` and `$NPM_VERSION` before applying the capture regex.

Verification:
- node --test test/workflow.test.mjs
- npm run prepack
- npm publish --dry-run --access public --provenance

**Comment by antongulin** [comment]
/review

Addressed latest non-inline concern in 0681081:
- Added static validation that rejects workflow guard scripts containing dynamic loading/execution tokens before the test helper executes the extracted script.

Verification:
- node --test test/workflow.test.mjs
- npm run prepack
- npm publish --dry-run --access public --provenance

File: plugin/test/workflow.test.mjs [hunk 7]
```diff
@@ -19,6 +41,9 @@ test("publish workflow uses npm trusted publishing provenance", () => {
...
```
[the remark this answers is not in the document]
Comment (antongulin) [reply plugin/test/workflow.test.mjs:41]: Not changing: this is already covered by the next test, `publish workflow rejects incomplete npm versions for trusted publishing`, which extracts and executes the same workflow script against invalid and valid npm versions. The provenance test remains a structural workflow assertion; the functional version-check behavior is verified separately.
